src/kbd-slide.c

```diff
@@ -0,0 +1,217 @@
+ * @return  TRUE if found, FALSE otherwise
+ */
+static gboolean
+kbd_slide_find_evdev()
+{
+    DIR *dir = NULL;
+    struct dirent *direntry = NULL;
+    gboolean result = FALSE;
+
+    if ((dir = opendir(DEV_INPUT_PATH)) == NULL) {
+        perror(__func__);
+        return FALSE;
+    }
+
+    for (direntry = readdir(dir);
+         (direntry != NULL && telldir(dir));
+         direntry = readdir(dir))
+    {
+        gchar *dev_path = NULL;
+        if (strncmp(direntry->d_name, EVENT_FILE_PREFIX,
+                    strlen(EVENT_FILE_PREFIX)) != 0)
+        {
+            continue;
+        }
+
+        dev_path = g_strconcat(DEV_INPUT_PATH, "/", direntry->d_name, NULL);
+
+        if (kbd_slide_set_input_dev(dev_path)) {
+            result = kbd_slide_add_handler();
+            g_free(dev_path);
+            break;
+        }
+
+        g_free(dev_path);
+    }
+
+    /* Report, but ignore, errors when closing directory */
+    if (closedir(dir) == -1)
+        perror(__func__);
+
+    return result;
+}
+
+/**
+ * Keypad slide state changes handler
+ */
+static gboolean
+kbd_slide_handler(GIOChannel *src, GIOCondition cond, gpointer data)
+{
+    struct input_event ev;
+    int rc;
+
+    for (;;) {
+        rc = libevdev_next_event(kbd_slide_dev, LIBEVDEV_READ_FLAG_NORMAL, &ev);
+
+        if (rc == LIBEVDEV_READ_STATUS_SYNC) {
+            /* We are out of sync, fix it */
+            do {
+                rc = libevdev_next_event(kbd_slide_dev,
+                                         LIBEVDEV_READ_FLAG_SYNC, &ev);
+            } while (rc == LIBEVDEV_READ_STATUS_SYNC);
+
+            /* Re-read current keyboard slide state */
+            gboolean state = libevdev_get_event_value(kbd_slide_dev,
+                                                      EV_SW, SW_KEYPAD_SLIDE);
+            inform_slide_keyboard(state);
+            continue;
+        }
+
+        if (rc == LIBEVDEV_READ_STATUS_SUCCESS) {
+            if (libevdev_event_is_code(&ev, EV_SW, SW_KEYPAD_SLIDE)) {
+                inform_slide_keyboard(ev.value);
+            }
+            continue;
+        }
+
+        if (rc == -EAGAIN) {
+            /* There's no more input data */
+            return TRUE;
+        }
+
+        /* An error occured or we are out of sync, stop handling */
+        return FALSE;
+    }
+}
+
+static gboolean
+kbd_slide_add_handler()
+{
+    kbd_slide_iochan = g_io_channel_unix_new(libevdev_get_fd(kbd_slide_dev));
+    if (kbd_slide_iochan == NULL)
+        return FALSE;
+
+    kbd_slide_src_id = g_io_add_watch(kbd_slide_iochan, G_IO_IN,
+                                      kbd_slide_handler, NULL);
+    return TRUE;
+}
+
+void
+kbd_slide_monitor_start()
+{
+    if (kbd_slide_dev != NULL) {
+        /* Already started */
+        return;
+    }
+
+    if (!kbd_slide_find_evdev()) {
+        /* Keypad slide state is considered closed if the device is not found */
+        inform_slide_keyboard(FALSE);
```

Comment:
This seems to differ from the original code, when in case slide_kbd_udi == NULL, inform_slide_keyboard(TRUE) will be called, unless I am missing something.

Reply:
Sure. You miss the change in src/events.c.
It's done in the same commit.

Reply:
Could you point me to the code that calls inform_slide_keyboard(TRUE) in case there is no suitable device found? I was not able to find it.

Reply:
I didn't find that code as well. It's you stated that this code exists.

Reply:
see https://github.com/maemo-leste/ke-recv/pull/2/commits/5f03b6827d550fd08ca9b7930d1329fa0297368a#diff-dce7b59c73d3da5554e4f44f417b19e4L660

This is what I meant by saying "original code"

This one calls inform_slide_keyboard(TRUE) if slide_kbd_udi == NULL, while you call inform_slide_keyboard(FALSE) in case there is no slide.

Reply:
So what's the problem? Did you check the change in src/events.c I mentioned above?

Reply:
Could you link me to that change? Sorry, cannot find it.

The problem is that you are changing the state from CLOSED to OPENED (or vice versa, not sure what TRUE/FALSE is in terms of slide state) when there is no slide.

Reply:
TRUE = the slide is present and is open
FALSE = slide is closed or there's no slide
I left a comment to the diff, you should receive a e-mail with the link, IIUC.